posthog/permissions.py

```diff
@@ -242,3 +246,106 @@ def has_permission(self, request, view) -> bool:
             return view.action in view.sharing_enabled_actions
 
         return False
+
+
+class APIScopePermission(BasePermission):
+    """
+    The request is via an API key and the user has the appropriate scopes.
+
+    This permission requires that the view has a "scope" attribute which is the base scope required for the action.
+    E.g. scope="insight" for a view that requires "insight:read" or "insight:write" for the relevant actions.
+
+    Actions can override this default scope by setting the `required_scopes` attribute on the view method.
+
+    """
+
+    write_actions = ["create", "update", "partial_update", "destroy"]
+    read_actions = ["list", "retrieve"]
+
+    def has_permission(self, request, view) -> bool:
+        # NOTE: We do this first to error out quickly if the view is missing the required attribute
+        # Helps devs remember to add it.
+        self.get_base_scope(request, view)
+
+        # API Scopes currently only apply to PersonalAPIKeyAuthentication
+        if not isinstance(request.successful_authenticator, PersonalAPIKeyAuthentication):
+            return True
+
+        key_scopes = request.successful_authenticator.personal_api_key.scopes_list
+
+        # TRICKY: Legacy API keys have no scopes and are allowed to do anything, even if the view is unsupported.
+        if not key_scopes:
+            return True
+
+        self.check_team_and_org_permissions(request, view)
+        required_scopes = self.get_required_scopes(request, view)
+
+        if "*" in key_scopes:
+            return True
+
+        for required_scope in required_scopes:
+            valid_scopes = [required_scope]
+
+            # For all valid scopes with :read we also add :write
+            for scope in valid_scopes:
+                if scope.endswith(":read"):
+                    valid_scopes.append(scope.replace(":read", ":write"))
+
+            if not any(scope in key_scopes for scope in valid_scopes):
+                raise PermissionDenied(f"API key missing required scope '{required_scope}'")
```

Comment:
AFAIK it would be more DRF-ic to do `self.message = "API key ..."` and then return false instead of throwing "manually".